lib/semantics/check-do-stmt.h

```diff
@@ -12,30 +12,21 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#ifndef FORTRAN_SEMANTICS_CHECK_DO_CONCURRENT_H_
-#define FORTRAN_SEMANTICS_CHECK_DO_CONCURRENT_H_
+#ifndef FORTRAN_SEMANTICS_CHECK_DO_STMT_H_
+#define FORTRAN_SEMANTICS_CHECK_DO_STMT_H_
 
 #include "semantics.h"
-#include "../common/indirection.h"
+#include "../parser/parse-tree.h"
 
-namespace Fortran::parser {
-struct DoConstruct;
-}
```

Comment:
Why replace the forward declaration with an include? In general it's good for `.h` files to include as little as possible (while still being self-contained). In this case `parse-tree.h` may be included wherever `check-do-stmt.h` is so it doesn't matter, but that's not obvious and it could change.

Reply:
I don't understand the underlying principle here.  The advantage of the include is that it tells me where DoConstruct is defined.

What's the advantage of making .h files include as little as possible?

Reply:
The advantage is potentially faster compilation. If a `.cc` file includes `check-do-concurrent.h` and not `parse-tree.h`, it will compile faster with the forward declaration because the compiler won't need to analyze `parse-tree.h`.

Reply:
Not a choice I would have made.  But I will now.